gvr-controller/app/src/main/java/org/gearvrf/sample/controller/SampleMain.java

```diff
@@ -173,162 +184,241 @@ public void onInit(GVRContext gvrContext) {
-                gvrCursorController.resetSceneObject();
+        public void onCursorControllerRemoved(GVRCursorController gvrCursorController)
+        {
+            if (gvrCursorController == mCursorController)
+            {
+                deselectController();
+            }
+            if (gvrCursorController.getControllerType() == GVRControllerType.CONTROLLER)
+            {
+                android.util.Log.d(TAG, "Removed remote orientation controller");
+            }
+            else if (gvrCursorController.getControllerType() == GVRControllerType.GAZE)
+            {
+                android.util.Log.d(TAG, "Removed gaze controller");
             }
         }
-    };
 
-    private GVRCursorController.ControllerEventListener controllerEventListener = new
-            GVRCursorController.ControllerEventListener() {
-        @Override
-        public void onEvent(GVRCursorController gvrCursorController) {
-            KeyEvent keyEvent = gvrCursorController.getKeyEvent();
-            if(keyEvent != null){
-                mPickHandler.setClicked(keyEvent.getAction() == KeyEvent.ACTION_DOWN);
-            }
+        public void onCursorControllerActive(GVRCursorController gvrCursorController)
+        {
+            onCursorControllerAdded(gvrCursorController);
         }
-    };
 
 
+        public void selectGearController(GVRCursorController controller)
+        {
+            GVRGearControllerSceneObject controllerModel = new GVRGearControllerSceneObject(mGVRContext);
+            controllerModel.setCursorController(controller);
+            controllerModel.setRayDepth(DEPTH);
+            controllerModel.setCursor(cursor);
+            controllerModel.enableSurfaceProjection();
+            controller.addPickEventListener(mPickHandler);
+            mControllerModel = controllerModel;
+            mCursorController = controller;
+            mGearController = (GearCursorController) controller;
+        }
 
-    private class PickHandler implements IPickEvents {
-        boolean clicked = false;
-        boolean prevClicked = false;
-        ButtonState state = ButtonState.UP;
 
-        public void setClicked(boolean clicked) {
-            prevClicked = this.clicked;
-            this.clicked = clicked;
+        public void selectGazeController(GVRCursorController controller)
+        {
+            GVRCursorControllerSceneObject controllerModel = new GVRCursorControllerSceneObject(mGVRContext);
+            mainScene.getMainCameraRig().getHeadTransformObject().addChildObject(controllerModel);
+            controllerModel.setCursorController(controller);
+            controllerModel.setRayDepth(DEPTH);
+            controllerModel.setCursor(cursor);
+            controllerModel.enableSurfaceProjection();
+            controller.addPickEventListener(mPickHandler);
+            mControllerModel = controllerModel;
+            mCursorController = controller;
+            mGazeController = (GVRGazeCursorController) controller;
```

Comment:
Why is the abstraction of the past removed? The users need to be aware of the specific instances of cursor controller classes now it seems. Presumably the users need to do different things depending on the type. We should be going into a direction of abstracting more than the old api did instead.

Reply:
This has always been true. This sample has never supported the Gaze controller before. The Gaze controller wasn't able to generate touch events by itself before my change. The app had to intercept them. There have always been a bunch of different cursor controller classes - I did not change these APIs.

Reply:
Then we should do correctly the gaze controller and keep it transparent to the end users.